minigalaxy/installer.py

```diff
@@ -349,12 +374,34 @@ def uninstall_game(game):
         os.remove(path_to_shortcut)
 
 
-def _exe_cmd(cmd):
-    process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-    stdout, stderr = process.communicate()
-    stdout = stdout.decode("utf-8")
-    stderr = stderr.decode("utf-8")
-    return stdout, stderr, process.returncode
+def _exe_cmd(cmd, capture_output=True, print_output=False):
+    """Wine commands are very verbose.
+    We should consume the output in regularly to prevent buffers (and minigalaxy memory) from filling up"""
+    print(f'executing wine command: {shlex.join(cmd)}')
+    std_out = []
+    process = subprocess.Popen(cmd,
+                               stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
+                               bufsize=1, universal_newlines=True, encoding="utf-8")
+    rc = process.poll()
+    while rc is None:
+        out_line = process.stdout.readline()
+        if capture_output and out_line != '':
+            std_out.append(out_line)
+
+        if print_output:
+            print(out_line, end='')
+
+        rc = process.poll()
+
+    print('command finished, read remaining output (if any)')
+    for line in process.stdout.readlines():
+        std_out.append(line)
+    print('done')
+
+    process.stdout.close()
+
+    output = ''.join(std_out)
+    return output, output, rc
```

Comment:
Was the old _exe_cmd function causing crashes?

Reply:
Not a crash, but a freeze or deadlock.
I tried to install 'lethis - path of progress' i think and it just didn't proceed in the post-install step that installs directx.
The problem with the old variant is that it just collected/buffered the entire process output first. But if the output of the process is bigger than the buffer, the process is blocked until it can write again and running directx install in wine generates a lot of output.
After i changed the _exe_cmd to consume buffers, the installation that failed before ran through without a problem.

Reply:
This is addressed by my PR: https://github.com/GB609/minigalaxy/pull/2